model/model_training/custom_datasets/formatting.py

```diff
@@ -15,20 +28,291 @@ def format_system_prefix(prefix, eos_token):
+    OM = "om"  # Oromo
+    OR = "or"  # Oriya
+    OS = "os"  # Ossetian, Ossetic
+    PA = "pa"  # Panjabi, Punjabi
+    PI = "pi"  # Pali
+    FA = "fa"  # Persian
+    PL = "pl"  # Polish
+    PS = "ps"  # Pashto, Pushto
+    PT = "pt"  # Portuguese
+    QU = "qu"  # Quechua
+    RM = "rm"  # Romansh
+    RN = "rn"  # Rundi
+    RO = "ro"  # Romanian, Moldavian, Moldovan
+    RU = "ru"  # Russian
+    SA = "sa"  # Sanskrit
+    SC = "sc"  # Sardinian
+    SD = "sd"  # Sindhi
+    SE = "se"  # Northern Sami
+    SM = "sm"  # Samoan
+    SG = "sg"  # Sango
+    SR = "sr"  # Serbian
+    GD = "gd"  # Gaelic, Scottish Gaelic
+    SN = "sn"  # Shona
+    SI = "si"  # Sinhala, Sinhalese
+    SK = "sk"  # Slovak
+    SL = "sl"  # Slovenian
+    SO = "so"  # Somali
+    ST = "st"  # Southern Sotho
+    ES = "es"  # Spanish, Castilian
+    SU = "su"  # Sundanese
+    SW = "sw"  # Swahili
+    SS = "ss"  # Swati
+    SV = "sv"  # Swedish
+    TA = "ta"  # Tamil
+    TE = "te"  # Telugu
+    TG = "tg"  # Tajik
+    TH = "th"  # Thai
+    TI = "ti"  # Tigrinya
+    BO = "bo"  # Tibetan
+    TK = "tk"  # Turkmen
+    TL = "tl"  # Tagalog
+    TN = "tn"  # Tswana
+    TO = "to"  # Tonga (Tonga Islands)
+    TR = "tr"  # Turkish
+    TS = "ts"  # Tsonga
+    TT = "tt"  # Tatar
+    TW = "tw"  # Twi
+    TY = "ty"  # Tahitian
+    UG = "ug"  # Uighur, Uyghur
+    UK = "uk"  # Ukrainian
+    UR = "ur"  # Urdu
+    UZ = "uz"  # Uzbek
+    VE = "ve"  # Venda
+    VI = "vi"  # Vietnamese
+    VO = "vo"  # Volap√ºk
+    WA = "wa"  # Walloon
+    CY = "cy"  # Welsh
+    WO = "wo"  # Wolof
+    FY = "fy"  # Western Frisian
+    XH = "xh"  # Xhosa
+    YI = "yi"  # Yiddish
+    YO = "yo"  # Yoruba
+    ZA = "za"  # Zhuang, Chuang
+    ZU = "zu"  # Zulu
+
+
+class DatasetEntry(BaseModel):
+    questions: list[str]
+    answers: list[str]
+    context: str | None
+    lang: Language | None
+    length: int | None
+    quality: float | None
+    humor: float | None
+    creativity: float | None
+
+    @validator("length")
+    def above_zero(cls, v) -> int:
+        if v is not None and v < 0:
+            raise ValueError(f"Length cannot be lower than 0. Received {v}")
+        return v
+
+    @validator("quality", "humor", "creativity")
+    def between_0_1(cls, v, field: ModelField) -> float:
+        if v is not None and not (0 <= v <= 1):
+            raise ValueError(f"Field {field.name} must be between 0 and 1. Received {v}.")
+        return v
+
+    def system_tag(self, eos_token: str) -> str | None:
+        relevant_system_infos = [
+            (k, v)
+            for k, v in self.__dict__.items()
+            if k not in ["questions", "answers"] and v is not None and v.replace("\n", "")
+        ]
+        if len(relevant_system_infos) > 0:
+            shuffle(relevant_system_infos)
+            system_tag_key_values = "\n".join([f"{k}: {v}" for k, v in relevant_system_infos])
```

Comment:
Should the values be quantized here or are they already within a certain set? (I think real values could be difficult to understand for the model)

Reply:
The keys here are one of `context`, `lang`, `length`, `quality`, `humor` and `creativity`. The values for `quality`, `humor` and `creativity` can be floating point numbers between 0 and 1 (this is checked by the validators and a test for this is added aswell, see [here](https://github.com/LAION-AI/Open-Assistant/pull/2520/files#diff-591736dab256f33df5fdc944907185433119c5c8032cdc40ada2beeba0a793d7R49)).

Reply:
yes, so if they can be anything between 0 and 1 (including e.g. 0.1826357216), then I think we should round them (e.g. to a single decimal) to make it easier for the model to understand.

Reply:
done